Fix grappler crash after DML device removal

PatriceVignola opened the pull request:
After a device removal, the adapter list will not necessarily be empty if the computer has more than one device, but it will contains at least 1 less element. Because of that, the device ID may become invalid.

adtsai commented on tensorflow/core/grappler/clusters/utils.cc:
```diff
@@ -160,13 +160,13 @@ DeviceProperties GetLocalDMLInfo(int device_id) {
+  // After a DML device gets removed, the device IDs in subsequent runs may be
+  // out of bounds since the list is smaller by 1 element
+  if (device_id >= adapters.size()) {
     return device;
```
How do we get into this state in the first place? The list of adapters is enumerated at startup, no?

PatriceVignola commented:
> How do we get into this state in the first place? The list of adapters is enumerated at startup, no? 

You're right; it's enumerated at startup, but grappler does its own enumeration, which gets the list of all devices minus the ones removed. This means that after a device removal, the user can try to use the index of the device that just got removed, which may be out of bounds.

I think a better solution here would be for grappler to know exactly which indices have been removed, otherwise the indices may not match the right device. Grappler shouldn't enumerate its own devices every time since it doesn't always match the devices that the user sees.

adtsai commented:
> > How do we get into this state in the first place? The list of adapters is enumerated at startup, no?
> 
> You're right; it's enumerated at startup, but grappler does its own enumeration, which gets the list of all devices minus the ones removed. This means that after a device removal, the user can try to use the index of the device that just got removed, which may be out of bounds.
> 
> I think a better solution here would be for grappler to know exactly which indices have been removed, otherwise the indices may not match the right device. Grappler shouldn't enumerate its own devices every time since it doesn't always match the devices that the user sees.

Yeah, if we consider device removal to be non-fatal then grappler's call to EnumerateAdapters is no longer correct. It really needs to be looking at the same list of adapters as the device factory. Otherwise, what happens if e.g. the 0th device gets removed in a dual-GPU system? Device ID 0 will still be valid, but will point to the wrong GPU.

PatriceVignola commented:
I extracted `DmlDeviceCache` into its own file and moved the `dml_runtime` rules into the `tensorflow/core` folder, where the other device runtimes (gpu_runtime and sycl_runtime) are located. This is needed to break a cycle in the dependency graph because dml_runtime used to depend on core_cpu. By moving dml_runtime 2 folders up, we can instead depend only the headers that we need without having to link against the CPU core library.

jstoecker reviewed (approved):
Looks good to me after the duplicate device registration is resolved for Linux build.

jstoecker commented on tensorflow/core/common_runtime/dml/dml_device_cache.h:
```diff
@@ -0,0 +1,50 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+==============================================================================*/
+
```
#pragma once

adtsai reviewed (approved):
:shipit:
